numpyro/contrib/module.py

```diff
@@ -433,3 +435,132 @@ def random_haiku_module(
         _update_params(params, new_params, prior)
     nn_new = partial(nn.func, new_params, *nn.args[1:], **nn.keywords)
     return nn_new
+
+
+def nnx_module(name, nn_module):
+    """
+    Declare a :mod:`~flax.nnx` style neural network inside a
+    model so that its parameters are registered for optimization via
+    :func:`~numpyro.primitives.param` statements.
+
+    Given a flax NNX ``nn_module``, to evaluate the module, we directly call it.
+    In a NumPyro model, the pattern will be::
+
+        # Eager initialization outside the model
+        module = nn_module(...)
+
+        # Inside the model
+        net = nnx_module("net", module)
+        y = net(x)
+
+    :param str name: name of the module to be registered.
+    :param flax.nnx.Module nn_module: a pre-initialized `flax nnx` Module instance.
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
+    try:
+        from flax import nnx
+    except ImportError as e:
+        raise ImportError(
+            "Looking like you want to use flax.nnx to declare "
+            "nn modules. This is an experimental feature. "
+            "You need to install the latest version of `flax` to use this feature. "
+            "It can be installed with `pip install git+https://github.com/google/flax.git`."
+        ) from e
+
+    graph_def, eager_params_state, eager_other_state = nnx.split(
+        nn_module, nnx.Param, nnx.Not(nnx.Param)
+    )
+
+    eager_params_state_dict = nnx.to_pure_dict(eager_params_state)
+
+    module_params = None
+    if eager_params_state:
+        module_params = numpyro.param(name + "$params")
+    if module_params is None:
+        module_params = numpyro.param(name + "$params", eager_params_state_dict)
+
+    eager_other_state_dict = nnx.to_pure_dict(eager_other_state)
+
+    mutable_holder = None
+    if eager_other_state_dict:
+        mutable_holder = numpyro_mutable(name + "$state")
+    if mutable_holder is None:
+        numpyro_mutable(name + "$state", {"state": eager_other_state_dict})
+
+    state_dict = {"state": eager_other_state}
+
+    def apply_fn(params, *call_args, **call_kwargs):
+        model = nn_module
+
+        if params:
+            graph_def, state = nnx.split(model)
+
+            if params:
```

Comment:
See the comment above - you don't need to split the model again
```
params_state = eager_params_state
if params:
    nnx.replace_by_pure_dict(params_state, params)
other_state = eager_other_state
if mutable_holder:
    nnx.replace_by_pure_dict(other_state, mutable_holder["state"])
model = nnx.merge(graph_def, params_state, other_state)
```

Reply:
You will need to use the latest mutable state, not just the latest params.

Reply:
mmm I am a bit confused: I tried this in https://github.com/pyro-ppl/numpyro/pull/1990/commits/17db8c14c20e393e98dd57edd700ce86d8217f36 and I get from the tests

```python
FAILED test/contrib/test_module.py::test_nnx_state_dropout_smoke[True-True] - TypeError: unsupported operand type(s) for *: 'float' and 'VariableState'
FAILED test/contrib/test_module.py::test_nnx_state_dropout_smoke[True-False] - TypeError: unsupported operand type(s) for *: 'float' and 'VariableState'
```

Is if ok that the `apply` function has access to parameters outside of it like `eager_params_state` and `mutable_holder`?

Reply:
I guess it is due to the line `    mutable_holder["state"] = eager_other_state` above. We can remove that line because we already have the logic to get the latest mutable state from the numpyro.mutable call.

Reply:
mmm I removed this line (https://github.com/pyro-ppl/numpyro/pull/1990/commits/5ffcfe30a35ffa6bc768f9aef7222e3bc5b55225) and keep seeing the tests failing. I wonder if there is something missing in https://github.com/pyro-ppl/numpyro/commit/17db8c14c20e393e98dd57edd700ce86d8217f36 🤔

Reply:
I wonder if we are updating the parameters more than needed in the apply. In https://github.com/pyro-ppl/numpyro/pull/1990/commits/12933f64a5680300577496967ea2b6252fae5993 I removed the last split and update just thinking we did the update just before calling the model. It seems the test with batch-normalization passes now. What do you think, or could the error be coming from somewhere else 🙏 ?

Reply:
I brought it back in https://github.com/pyro-ppl/numpyro/pull/1990/commits/d25de60929d3659aa1449706b3dc8eba1a0e925d because you have been pointing out this in many times... I can't figure out the error in the tests ...